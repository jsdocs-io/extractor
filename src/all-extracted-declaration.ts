import type {
  ExtractedClass,
  ExtractedClassConstructor,
  ExtractedClassMethod,
  ExtractedClassProperty,
} from "./extract-class";
import type { ExtractedEnum, ExtractedEnumMember } from "./extract-enum";
import type { ExtractedFunction } from "./extract-function";
import type {
  ExtractedInterface,
  ExtractedInterfaceCallSignature,
  ExtractedInterfaceConstructSignature,
  ExtractedInterfaceGetAccessor,
  ExtractedInterfaceIndexSignature,
  ExtractedInterfaceMethod,
  ExtractedInterfaceProperty,
  ExtractedInterfaceSetAccessor,
} from "./extract-interface";
import type { ExtractedNamespace } from "./extract-namespace";
import type { ExtractedTypeAlias } from "./extract-type-alias";
import type { ExtractedVariable } from "./extract-variable";

/**
`AllExtractedDeclaration` is the union of all possible declarations
that can be extracted, with some being found only in other declarations
(e.g., class method declarations are found only in a class declaration).
*/
export type AllExtractedDeclaration =
  | ExtractedVariable
  | ExtractedFunction
  | ExtractedClass
  | ExtractedClassConstructor
  | ExtractedClassProperty
  | ExtractedClassMethod
  | ExtractedInterface
  | ExtractedInterfaceProperty
  | ExtractedInterfaceMethod
  | ExtractedInterfaceConstructSignature
  | ExtractedInterfaceCallSignature
  | ExtractedInterfaceIndexSignature
  | ExtractedInterfaceGetAccessor
  | ExtractedInterfaceSetAccessor
  | ExtractedEnum
  | ExtractedEnumMember
  | ExtractedTypeAlias
  | ExtractedNamespace;

/**
`AllExtractedDeclarationKind` is the union of all discriminators
used to detect the kind of declaration.
*/
export type AllExtractedDeclarationKind = AllExtractedDeclaration["kind"];
