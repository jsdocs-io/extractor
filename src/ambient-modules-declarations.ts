import { ModuleDeclaration, Node, type Project } from "ts-morph";
import { isHidden } from "./is-hidden";
import { sourceFilePath } from "./source-file-path";

export type AmbientModulesDeclarationsReturn = {
  containerName: string;
  exportName: string;
  declaration: ModuleDeclaration;
}[];

export const ambientModulesDeclarations = (
  containerName: string,
  project: Project,
  pkgName?: string,
): AmbientModulesDeclarationsReturn => {
  const ambientModulesDeclarations = [];
  for (const symbol of project.getAmbientModules()) {
    for (const declaration of symbol.getDeclarations()) {
      if (isHidden(declaration) || !Node.isModuleDeclaration(declaration)) {
        continue;
      }
      if (pkgName && !sourceFilePath(declaration).startsWith(`/${pkgName}`)) {
        // Ignore ambient modules that are not from the analyzed package.
        continue;
      }
      const exportName = declaration.getName();
      // TODO:
      // Remove surrounding quotes and eventual spaces
      // const exportID = exportName
      //   .replace(/"|'/g, "")
      //   .replace(/\s/g, "_")
      //   .trim();
      ambientModulesDeclarations.push({
        containerName,
        exportName,
        declaration,
      });
    }
  }
  return ambientModulesDeclarations;
};
