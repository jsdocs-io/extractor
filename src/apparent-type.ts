import { TypeFormatFlags, ts, type Node } from "ts-morph";

export function apparentType(node: Node): string {
	// See:
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-427405736
	// https://github.com/dsherret/ts-morph/issues/453#issuecomment-667578386
	return node
		.getType()
		.getApparentType()
		.getText(
			node,
			ts.TypeFormatFlags.NoTruncation | TypeFormatFlags.UseAliasDefinedOutsideCurrentScope,
		)
		.replace(/^Number$/, "number")
		.replace(/^Boolean$/, "boolean")
		.replace(/^String$/, "string");
}
