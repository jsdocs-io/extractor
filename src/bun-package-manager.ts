import { Effect } from "effect";
import { execa } from "execa";
import { InstallPackageError, PackageManager } from "./package-manager";

/** @internal */
export const bunPackageManager = (bunPath = "bun") =>
	PackageManager.of({
		installPackage: ({ pkg, cwd }) =>
			Effect.gen(function* (_) {
				// Run `bun add <pkg> --verbose`.
				// See https://bun.sh/docs/cli/add.
				const { stdout } = yield* _(
					Effect.tryPromise({
						try: () => execa(bunPath, ["add", pkg, "--verbose"], { cwd }),
						catch: (e) => new InstallPackageError({ cause: e }),
					}),
				);

				// With verbose output on, bun prints one line per installed package
				// (e.g., "foo@1.0.0"), including all installed dependencies.
				// These lines are between the two delimiting lines found here:
				// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
				const lines = stdout.split("\n");
				const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
				const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
				const installedPackages = lines.slice(beginHash + 1, endHash);
				return installedPackages;
			}),
	});
