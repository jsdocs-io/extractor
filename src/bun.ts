import { execa } from "execa";

/** `Bun` is a wrapper for the Bun package manager. */
export class Bun {
	#bunCmd: string;

	constructor(bunCmd = "bun") {
		this.#bunCmd = bunCmd.trim();
	}

	/**
	`add` installs the given package in the given directory and
	returns a list of all installed packages (e.g., `["foo@1.0.0"]`).

	@see {@link https://bun.sh/docs/cli/add}
	*/
	async add(pkg: string, cwd: string): Promise<string[]> {
		// Run `bun add <pkg> --verbose` in the `cwd`.
		const { stdout } = await execa(this.#bunCmd, ["add", pkg, "--verbose"], { cwd });

		// With verbose output on, bun prints one line per installed package
		// (e.g., "foo@1.0.0"), including all installed dependencies.
		// These lines are between the two delimiting lines found here:
		// https://github.com/oven-sh/bun/blob/972a7b7080bd3066b54dcb43e9c91c5dfa26a69c/src/install/lockfile.zig#L5369-L5370.
		const lines = stdout.split("\n");
		const beginHash = lines.findIndex((line) => line.startsWith("-- BEGIN SHA512/256"));
		const endHash = lines.findIndex((line) => line.startsWith("-- END HASH"));
		const installedPackages = lines.slice(beginHash + 1, endHash);
		return installedPackages;
	}
}
