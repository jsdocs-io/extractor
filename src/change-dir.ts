import { Result } from "neverthrow";
import { OsError } from "./errors";

export const changeDir = Result.fromThrowable(
  process.chdir,
  (e) => new OsError("failed to change directory", { cause: e }),
);
