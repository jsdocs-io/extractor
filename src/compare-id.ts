export type WithId = {
  id: string;
};

export const compareId = (a: WithId, b: WithId): number =>
  a.id.localeCompare(b.id);
