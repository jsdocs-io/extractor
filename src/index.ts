export {
  extractDeclarations,
  type ExtractDeclarationsOptions,
  type ExtractedDeclaration,
  type ExtractedDeclarationKind,
} from "./extract-declarations";
