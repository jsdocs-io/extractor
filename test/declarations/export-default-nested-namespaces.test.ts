import dedent from "ts-dedent";
import {
  ModuleKind,
  ModuleResolutionKind,
  Project,
  ScriptTarget,
} from "ts-morph";
import { expect, test } from "vitest";
import { extractDeclarations } from "../../src";

test("export default nested namespaces", async () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      lib: ["lib.esnext.full.d.ts"],
      target: ScriptTarget.ESNext,
      module: ModuleKind.ESNext,
      moduleResolution: ModuleResolutionKind.Bundler,
    },
  });
  const indexFile = project.createSourceFile(
    "index.d.ts",
    dedent`
    /**
     * This is similar to the structure of \`index.d.ts\` for \`firebase@8.2.2\`.
     *
     * @packageDocumentation
     */

    /** foo doc */
    declare namespace foo {
      /** fooFunc doc */
      function fooFunc(s: string): string;
    }

    /** foo.bar doc */
    declare namespace foo.bar {
      /** foobarFunc doc */
      function foobarFunc(n: number): number;
    }

    /** foo.bar.baz doc */
    declare namespace foo.bar.baz {
      /** foobarbazFunc doc */
      function foobarbazFunc(b: boolean): boolean;
    }

    // Namespace \`abc\` is at depth 6 and should not be extracted
    declare namespace foo.bar.baz.qux.xyz.abc {}

    export default foo;
    `,
  );
  expect(
    await extractDeclarations({
      containerName: "",
      container: indexFile,
      maxDepth: 5,
      project,
    }),
  ).toMatchSnapshot();
});
