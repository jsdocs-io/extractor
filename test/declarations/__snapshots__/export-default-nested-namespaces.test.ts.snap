// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`export default nested namespaces 1`] = `
[
  {
    "declarations": [
      {
        "docs": [
          "/** fooFunc doc */",
        ],
        "file": "/index.d.ts",
        "id": "namespace.default.function.fooFunc",
        "kind": "function",
        "line": 10,
        "name": "fooFunc",
        "signature": "fooFunc: (s: string) => string;",
      },
      {
        "declarations": [
          {
            "docs": [
              "/** foobarFunc doc */",
            ],
            "file": "/index.d.ts",
            "id": "namespace.default.namespace.bar.function.foobarFunc",
            "kind": "function",
            "line": 16,
            "name": "foobarFunc",
            "signature": "foobarFunc: (n: number) => number;",
          },
          {
            "declarations": [
              {
                "docs": [
                  "/** foobarbazFunc doc */",
                ],
                "file": "/index.d.ts",
                "id": "namespace.default.namespace.bar.namespace.baz.function.foobarbazFunc",
                "kind": "function",
                "line": 22,
                "name": "foobarbazFunc",
                "signature": "foobarbazFunc: (b: boolean) => boolean;",
              },
              {
                "declarations": [
                  {
                    "declarations": [],
                    "docs": [],
                    "file": "/index.d.ts",
                    "id": "namespace.default.namespace.bar.namespace.baz.namespace.qux.namespace.xyz",
                    "kind": "namespace",
                    "line": 26,
                    "name": "xyz",
                    "signature": "namespace xyz {}",
                  },
                ],
                "docs": [],
                "file": "/index.d.ts",
                "id": "namespace.default.namespace.bar.namespace.baz.namespace.qux",
                "kind": "namespace",
                "line": 26,
                "name": "qux",
                "signature": "namespace qux {}",
              },
            ],
            "docs": [],
            "file": "/index.d.ts",
            "id": "namespace.default.namespace.bar.namespace.baz",
            "kind": "namespace",
            "line": 20,
            "name": "baz",
            "signature": "namespace baz {}",
          },
        ],
        "docs": [],
        "file": "/index.d.ts",
        "id": "namespace.default.namespace.bar",
        "kind": "namespace",
        "line": 14,
        "name": "bar",
        "signature": "namespace bar {}",
      },
    ],
    "docs": [
      "/** foo doc */",
    ],
    "file": "/index.d.ts",
    "id": "namespace.default",
    "kind": "namespace",
    "line": 8,
    "name": "default",
    "signature": "namespace default {}",
  },
]
`;
